File: core/src/de/fourtytwoways/onion/domain/usecases/contract/ContractDurationChange.java
package de.fourtytwoways.onion.domain.usecases.contract;
// (c) 2022 Thomas Herrmann, 42ways GmbH

import de.fourtytwoways.onion.domain.entities.contract.Contract;

import java.time.LocalDate;

public class ContractDurationChange {
    public static Contract adjustStartDate(Contract contract, LocalDate newStartDate, boolean calculateBenefit) {
        contract.setStartDate(newStartDate);
        if (calculateBenefit)
            return ContractCalculation.calculateBenefit(contract);
        else
            return ContractCalculation.calculatePremium(contract);
    }

    // unfortunately Java doesn't have default parameters, so we have to use this noisy code repetition
    public static Contract adjustStartDate(Contract contract, LocalDate newStartDate) {
        return adjustStartDate(contract, newStartDate, true);
    }

    public static Contract adjustEndDate(Contract contract, LocalDate newEndDate, boolean calculateBenefit) {
        contract.setEndDate(newEndDate);
        if (calculateBenefit)
            return ContractCalculation.calculateBenefit(contract);
        else
            return ContractCalculation.calculatePremium(contract);
    }

    // unfortunately Java doesn't have default parameters, so we have to use this noisy code repetition
    public static Contract adjustEndDate(Contract contract, LocalDate newEndDate) {
        return adjustEndDate(contract, newEndDate, true);
    }
}

File: core/src/main/java/de/fourtytwoways/onion/domain/usecases/contract/ContractDurationChange.kt
package de.fourtytwoways.onion.domain.usecases.contract
// (c) 2022 Thomas Herrmann, 42ways GmbH

import de.fourtytwoways.onion.domain.entities.contract.Contract
import de.fourtytwoways.onion.domain.values.enumeration.ComputationTarget
import java.time.LocalDate

class ContractDurationChange {
    @JvmOverloads
    fun adjustStartDate(contract: Contract, newStartDate: LocalDate, computationTarget: ComputationTarget = ComputationTarget.BENEFIT): Contract {
        // TODO: Wouldn't it be good to have contract immutable?
        contract.startDate = newStartDate
        return ContractCalculation().calculate(contract, computationTarget)
    }

    @JvmOverloads
    fun adjustEndDate(contract: Contract, newEndDate: LocalDate, computationTarget: ComputationTarget = ComputationTarget.BENEFIT): Contract {
        contract.endDate = newEndDate
        return ContractCalculation().calculate(contract, computationTarget)
    }
}